tty/imx6ull-uart/imx6ull-uart.c

```diff
@@ -77,21 +95,143 @@ typedef struct {
 	volatile uint32_t *base;
 	uint32_t mode;
 	uint16_t dev_no;
+	uint8_t busy;
 
 	handle_t cond;
 	handle_t inth;
 	handle_t lock;
+	handle_t openclose_lock;
+
+	uint8_t int_flags;
+
+	struct {
+		uint32_t parity;
+		uint32_t frame;
+		uint32_t hw_overrun;
+		uint32_t sw_overrun;
+	} counters;
+
+	lf_fifo_t *rx_sw_fifo; /* unprocessed characters read from HW FIFO */
 
 	libtty_common_t tty_common;
 } uart_t;
 
 uart_t uart = { 0 };
 
+#define READER 1
+#define WRITER 2
+
 #define MODULE_CLK 20000000
 
 #define BUFSIZE 4096
 
-void uart_thr(void *arg)
+#define RX_SW_FIFO_SIZE 256
+
+/* interrupt flags */
+#define RX_DONE        (1 << 0)
+#define TX_DONE        (1 << 1)
+#define PARITY_ERR     (1 << 2)
+#define FRAME_ERR      (1 << 3)
+#define HW_OVERRUN_ERR (1 << 4)
+#define SW_OVERRUN_ERR (1 << 5)
+
+/* register flags */
```

Comment:
IMHO more readable than magic numbers but adding @HBuczynski to reviewers for kernel team to approve the style :P.